out-variant for torch.batch_norm_elemt

vadimkantorov opened the pull request:
Following dicussion with @ezyang in https://github.com/pytorch/pytorch/issues/26288

vadimkantorov committed:
out-variant for torch.batch_norm_elemt

Tensor -> Tensor& as out-variant return type

batch_norm_elemt_cuda_template: reshape -> view for output

batch_norm_elemt_cuda_template return type: Tensor& -> void

vadimkantorov commented:
yes, compilation error, the same is I had got locally

ezyang commented on aten/src/ATen/native/cuda/Normalization.cu:
```diff
@@ -60,22 +60,28 @@ std::tuple<Tensor, Tensor> batch_norm_stats_cuda(const Tensor& self, double epsi
+  return batch_norm_elemt_cuda_out(output, self, weight, bias, mean, invstd, epsilon);
+}
+
+Tensor batch_norm_elemt_cuda_out(Tensor& output, const Tensor& self, const Tensor& weight, const Tensor& bias,
```
Return a Tensor& here and I think your problem will go away. You might need to adjust your inner code to directly return output.

vadimkantorov replied:
Is returning a reference to a local variable (tensor) safe?

vadimkantorov replied:
I'm not sure I'm understanding your suggestion. Do you mean replacing `Tensor` as the return type of `batch_norm_elemt_cuda_out` and `batch_norm_elemt_cuda` by `Tensor&`?

vadimkantorov replied:
Even if we change return type, in theory, the error should persist: `Oct 09 19:39:07 /var/lib/jenkins/workspace/aten/src/ATen/native/cuda/Normalization.cu(67): error: cannot overload functions distinguished by return type alone
`

ezyang replied:
> Is returning a reference to a local variable (tensor) safe?

No it is not. However, the convention is that `_out` functions return the reference for out that was passed in. It should look something like:

```
Tensor& foo_out(Tensor& output, const Tensor& input) {
  // ...
  return output;
}
```

ezyang replied:
> Even if we change return type, in theory, the error should persist

I'm not sure I follow. There might be other errors in your code, but your error message is consistent with this C++ code:

```
// Auto-generated in NativeFunctions.h
Tensor& foo_out(Tensor& output, const Tensor& input);

// Your implementation
Tensor foo_out(Tensor& output, const Tensor& input) {
  ...
}
```

This is exactly the situation a C++ compiler will complain about an overload distinguished only in return type.

By the way, you can find out exactly what signature ATen is expecting by inspecting the generated native functions header.

vadimkantorov replied:
@ezyang Thanks for explanation! I now understand you, I'll try the suggested return type for the `_out`-variant.

vadimkantorov commented:
clang-tidy check at https://github.com/pytorch/pytorch/pull/27621/checks?check_run_id=257692310 seems unrelated

and [compilation fails](https://circleci.com/gh/pytorch/pytorch/3182638?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-linkb) with:
```
HEAD is now at 3e7b27a2be Tensor -> Tensor& as out-variant return type
+ git merge --no-edit --no-ff fcb6dd079eb01b594aeea9f7c83ea1b36aa65793
fatal: refusing to merge unrelated histories
Exited with code 128
```
Do I need to rebase?

Also, https://github.com/vadimkantorov/pytorch/blob/master/aten/src/ATen/native/cuda/Normalization.cuh/#L678-L679 in an extreme case could reallocate, and then returning a reference to `output_` (before reshape) is incorrect. In practice this should never happen. Is there a simple way to ensure that reshape does not reallocate?

ezyang commented:
> Do I need to rebase? 

Yep.

> Also, https://github.com/vadimkantorov/pytorch/blob/master/aten/src/ATen/native/cuda/Normalization.cuh/#L678-L679 in an extreme case could reallocate, and then returning a reference to output_ (before reshape) is incorrect

You should error in that case. Replace this with a `view` which is guaranteed not to reallcoate?

vadimkantorov commented:
> You should error in that case. Replace this with a `view` which is guaranteed not to reallcoate?
Ok, I'll replace the `output_.reshape` with `output.view`. On a side note, currently there is no easy way for:
```python
a = torch.rand(2,3,4)
a_ = a.transpose(-1, -2)
b_ = someltwisefunc(a_.reshape(2, -1)) # reshape seems to reallocate, view will error out
b = b_.view_as(a)
```

vadimkantorov commented:
Squashed and rebased. I hope it's done correctly (I added remote upstream, fetched it, squashed commits and then rebased, it's my first time using `git rebase`, so sorry for this verbosity).

If compilation works, I'll replace `output_.reshape` to `output_.view`

ezyang commented:
Build failure is real

```

Oct 15 13:15:06 /var/lib/jenkins/workspace/aten/src/ATen/native/cuda/Normalization.cu(70): error: initial value of reference to non-const must be an lvalue
Oct 15 13:15:06 
```

ezyang commented on aten/src/ATen/native/cuda/Normalization.cuh:
```diff
@@ -669,14 +669,14 @@ std::tuple<Tensor, Tensor> batch_norm_stats_cuda_template(const Tensor& input_,
   int64_t n_input = input_.size(1);
   auto input_reshaped = input_.reshape({input_.size(0), input_.size(1), -1}); // internally we merge the feature dimensions
-  auto output_reshaped = at::empty_like(input_reshaped);
+  auto output_reshaped = output_.reshape({input_.size(0), input_.size(1), -1});
```
`view` here, to avoid realloc

ezyang commented:
The rest of the code looks reasonable!

vadimkantorov commented:
> Build failure is real
> 
> ```
> 
> Oct 15 13:15:06 /var/lib/jenkins/workspace/aten/src/ATen/native/cuda/Normalization.cu(70): error: initial value of reference to non-const must be an lvalue
> Oct 15 13:15:06 
> ```
Is it because `AT_DISPATCH_FLOATING_TYPES_AND_HALF` and the lambda passed in are implicitly returning a `Tensor`, not `Tensor&`? Would you have an advice how to fix that?

vadimkantorov commented on aten/src/ATen/native/cuda/Normalization.cu:
```diff
@@ -60,22 +60,29 @@ std::tuple<Tensor, Tensor> batch_norm_stats_cuda(const Tensor& self, double epsi
+          return batch_norm_elemt_cuda_template<at::Half, float, int64_t>(output, self, weight, bias, mean, invstd, epsilon);
         } else {
-          return batch_norm_elemt_cuda_template<scalar_t, scalar_t, int64_t>(self, weight, bias, mean, invstd, epsilon);
+          return batch_norm_elemt_cuda_template<scalar_t, scalar_t, int64_t>(output, self, weight, bias, mean, invstd, epsilon);
```
One way would be to remove `return`s here and replace them by `return output` at the end. Should I do that @ezyang?

ezyang replied:
Yeah, that's what I would do.

vadimkantorov commented:
finally it compiled :)
